File: src/lib/ebike/assistLog.ts
/**
 * F-EBK-005 — Assist-Modus-Protokollierung (P2)
 *
 * Nur Logging — keine Motorsteuerung (F-EBK-000).
 * Ohne OEM-Datenquelle: Schätzung aus Leistungs-/Geschwindigkeitssignatur
 * ODER manuelle Angabe — Schätzung MUSS als Schätzung gekennzeichnet sein.
 *
 * Quellen Verbrauchsfaktoren (Bosch Battery Guide / BIKE Magazin / myvelo):
 * Eco effizient, Tour gleichmäßig, Sport/eMTB progressiv, Turbo max. Support
 * → typisch 6–15 Wh/km, Turbo 2–3× Eco.
 */

export type AssistMode = "off" | "eco" | "tour" | "sport" | "turbo";

export interface AssistSegment {
  id: string;
  mode: AssistMode;
  /** 'oem' = ausgelesen, 'estimated' = Signatur, 'manual' = Nutzer */
  source: "oem" | "estimated" | "manual";
  startOffsetSec: number;
  endOffsetSec: number;
  distanceM: number;
  avgSpeedKmh: number;
  avgRiderPowerW?: number;
  estimatedWh?: number;
  label: string;
}

export interface AssistRideSummary {
  segments: AssistSegment[];
  dominantMode: AssistMode;
  modeSharePct: Record<AssistMode, number>;
  estimatedTotalWh: number;
  hasEstimates: boolean;
  disclaimer: string;
  sourceLabel: string;
}

/** Relative Wh/km-Faktoren vs. Tour=1 (Bosch/BIKE-Praxis, grob) */
export const ASSIST_WH_FACTOR: Record<AssistMode, number> = {
  off: 0.15,
  eco: 0.65,
  tour: 1.0,
  sport: 1.45,
  turbo: 2.1,
};

export function estimateModeFromSignature(input: {
  speedKmh: number;
  riderPowerW: number;
  gradeApprox?: number;
}): { mode: AssistMode; confidence: number } {
  const grade = input.gradeApprox ?? 0.03;
  const power = input.riderPowerW;
  const speed = input.speedKmh;

  // Hohe Geschwindigkeit bergauf bei niedriger Fahrerleistung → Turbo/Sport
  if (grade > 0.06 && speed > 14 && power < 120) {
    return { mode: "turbo", confidence: 0.55 };
  }
  if (grade > 0.04 && speed > 12 && power < 140) {
    return { mode: "sport", confidence: 0.5 };
  }
  if (speed < 10 && power > 160) {
    return { mode: "eco", confidence: 0.45 };
  }
  if (power > 180 && speed < 18) {
    return { mode: "eco", confidence: 0.4 };
  }
  return { mode: "tour", confidence: 0.35 };
}

/** Demo: segmente einen Ride in Assist-Abschnitte (Schätzung) */
export function buildEstimatedAssistLog(input: {
  durationSec: number;
  distanceM: number;
  elevationGainM: number;
  avgRiderPower?: number;
  avgSpeedKmh?: number;
  preferredMode?: AssistMode;
}): AssistRideSummary {
  const duration = Math.max(60, input.durationSec);
  const dist = Math.max(100, input.distanceM);
  const speed = input.avgSpeedKmh ?? (dist / 1000 / (duration / 3600));
  const power = input.avgRiderPower ?? 110;
  const grade = input.elevationGainM / Math.max(1, dist);

  const climbShare = Math.min(0.55, 0.2 + grade * 8);
  const segments: AssistSegment[] = [];

  // Segment 1: Anfahrt Tour/Eco
  const t1 = Math.round(duration * 0.25);
  const d1 = Math.round(dist * 0.28);
  const m1 = input.preferredMode === "eco" ? "eco" : "tour";
  segments.push({
    id: "seg-1",
    mode: m1,
    source: "estimated",
    startOffsetSec: 0,
    endOffsetSec: t1,
    distanceM: d1,
    avgSpeedKmh: speed * 0.95,
    avgRiderPowerW: power + 10,
    estimatedWh: (d1 / 1000) * 10 * ASSIST_WH_FACTOR[m1],
    label: `Schätzung: ${m1.toUpperCase()} (Anfahrt)`,
  });

  // Segment 2: Steigung → höherer Modus
  const climbEst = estimateModeFromSignature({
    speedKmh: speed * 0.7,
    riderPowerW: power * 0.8,
    gradeApprox: Math.max(grade, 0.07),
  });
  const t2 = Math.round(duration * climbShare);
  const d2 = Math.round(dist * 0.35);
  segments.push({
    id: "seg-2",
    mode: climbEst.mode,
    source: "estimated",
    startOffsetSec: t1,
    endOffsetSec: t1 + t2,
    distanceM: d2,
    avgSpeedKmh: speed * 0.75,
    avgRiderPowerW: power * 0.85,
    estimatedWh: (d2 / 1000) * 14 * ASSIST_WH_FACTOR[climbEst.mode],
    label: `Schätzung: ${climbEst.mode.toUpperCase()} (Steigung, Konfidenz ${(climbEst.confidence * 100).toFixed(0)} %)`,
  });

  // Segment 3: Rest Tour
  const t3 = duration - t1 - t2;
  const d3 = dist - d1 - d2;
  segments.push({
    id: "seg-3",
    mode: "tour",
    source: "estimated",
    startOffsetSec: t1 + t2,
    endOffsetSec: duration,
    distanceM: Math.max(0, d3),
    avgSpeedKmh: speed,
    avgRiderPowerW: power,
    estimatedWh: (Math.max(0, d3) / 1000) * 9 * ASSIST_WH_FACTOR.tour,
    label: "Schätzung: TOUR (Rest)",
  });

  return summarizeAssist(segments);
}

export function summarizeAssist(segments: AssistSegment[]): AssistRideSummary {
  const totals: Record<AssistMode, number> = {
    off: 0,
    eco: 0,
    tour: 0,
    sport: 0,
    turbo: 0,
  };
  let totalDist = 0;
  let wh = 0;
  for (const s of segments) {
    totals[s.mode] += s.distanceM;
    totalDist += s.distanceM;
    wh += s.estimatedWh ?? 0;
  }
  const share = {} as Record<AssistMode, number>;
  (Object.keys(totals) as AssistMode[]).forEach((m) => {
    share[m] = totalDist > 0 ? Math.round((totals[m] / totalDist) * 100) : 0;
  });
  const dominant = (Object.entries(totals).sort((a, b) => b[1] - a[1])[0]?.[0] ??
    "tour") as AssistMode;

  return {
    segments,
    dominantMode: dominant,
    modeSharePct: share,
    estimatedTotalWh: Math.round(wh),
    hasEstimates: segments.some((s) => s.source === "estimated"),
    disclaimer:
      "Schätzungen aus Leistungs-/Geschwindigkeitssignatur — kein OEM-Auslesen. Keine Motorsteuerung (F-EBK-000).",
    sourceLabel:
      "Bosch Battery Guide Moduswirkung · BIKE Magazin Reichweitentest · Spec F-EBK-005",
  };
}

export function manualAssistSegment(
  mode: AssistMode,
  durationSec: number,
  distanceM: number
): AssistSegment {
  return {
    id: `manual-${mode}`,
    mode,
    source: "manual",
    startOffsetSec: 0,
    endOffsetSec: durationSec,
    distanceM,
    avgSpeedKmh: distanceM / 1000 / (durationSec / 3600 || 1),
    estimatedWh: (distanceM / 1000) * 11 * ASSIST_WH_FACTOR[mode],
    label: `Manuell: ${mode.toUpperCase()}`,
  };
}

File: mobile/lib/domain/ebike/assist_log.dart
/// F-EBK-005 Assist-Modus-Protokollierung (Port assistLog.ts) — nur Schätzung.
library;

typedef AssistMode = String; // off | eco | tour | sport | turbo

class AssistSegment {
  const AssistSegment({
    required this.id,
    required this.mode,
    required this.source,
    required this.startOffsetSec,
    required this.endOffsetSec,
    required this.distanceM,
    required this.avgSpeedKmh,
    required this.label,
    this.avgRiderPowerW,
    this.estimatedWh,
  });

  final String id;
  final AssistMode mode;
  final String source; // oem | estimated | manual
  final int startOffsetSec;
  final int endOffsetSec;
  final double distanceM;
  final double avgSpeedKmh;
  final double? avgRiderPowerW;
  final double? estimatedWh;
  final String label;
}

class AssistRideSummary {
  const AssistRideSummary({
    required this.segments,
    required this.dominantMode,
    required this.modeSharePct,
    required this.estimatedTotalWh,
    required this.hasEstimates,
    required this.disclaimer,
    required this.sourceLabel,
  });

  final List<AssistSegment> segments;
  final AssistMode dominantMode;
  final Map<AssistMode, int> modeSharePct;
  final int estimatedTotalWh;
  final bool hasEstimates;
  final String disclaimer;
  final String sourceLabel;
}

const assistWhFactor = <AssistMode, double>{
  'off': 0.15,
  'eco': 0.65,
  'tour': 1.0,
  'sport': 1.45,
  'turbo': 2.1,
};

({AssistMode mode, double confidence}) estimateModeFromSignature({
  required double speedKmh,
  required double riderPowerW,
  double gradeApprox = 0.03,
}) {
  if (gradeApprox > 0.06 && speedKmh > 14 && riderPowerW < 120) {
    return (mode: 'turbo', confidence: 0.55);
  }
  if (gradeApprox > 0.04 && speedKmh > 12 && riderPowerW < 140) {
    return (mode: 'sport', confidence: 0.5);
  }
  if (speedKmh < 10 && riderPowerW > 160) {
    return (mode: 'eco', confidence: 0.45);
  }
  if (riderPowerW > 180 && speedKmh < 18) {
    return (mode: 'eco', confidence: 0.4);
  }
  return (mode: 'tour', confidence: 0.35);
}

AssistRideSummary buildEstimatedAssistLog({
  required int durationSec,
  required double distanceM,
  required double elevationGainM,
  double? avgRiderPower,
  double? avgSpeedKmh,
  AssistMode? preferredMode,
}) {
  final duration = durationSec < 60 ? 60 : durationSec;
  final dist = distanceM < 100 ? 100.0 : distanceM;
  final speed =
      avgSpeedKmh ?? (dist / 1000 / (duration / 3600));
  final power = avgRiderPower ?? 110;
  final grade = elevationGainM / (dist < 1 ? 1 : dist);
  final climbShare = (0.2 + grade * 8).clamp(0.0, 0.55);

  final segments = <AssistSegment>[];
  final t1 = (duration * 0.25).round();
  final d1 = (dist * 0.28).roundToDouble();
  final m1 = preferredMode == 'eco' ? 'eco' : 'tour';
  segments.add(
    AssistSegment(
      id: 'seg-1',
      mode: m1,
      source: 'estimated',
      startOffsetSec: 0,
      endOffsetSec: t1,
      distanceM: d1,
      avgSpeedKmh: speed * 0.95,
      avgRiderPowerW: power + 10,
      estimatedWh: (d1 / 1000) * 10 * (assistWhFactor[m1] ?? 1),
      label: 'Schätzung: ${m1.toUpperCase()} (Anfahrt)',
    ),
  );

  final climbEst = estimateModeFromSignature(
    speedKmh: speed * 0.7,
    riderPowerW: power * 0.8,
    gradeApprox: grade > 0.07 ? grade : 0.07,
  );
  final t2 = (duration * climbShare).round();
  final d2 = (dist * 0.35).roundToDouble();
  segments.add(
    AssistSegment(
      id: 'seg-2',
      mode: climbEst.mode,
      source: 'estimated',
      startOffsetSec: t1,
      endOffsetSec: t1 + t2,
      distanceM: d2,
      avgSpeedKmh: speed * 0.75,
      avgRiderPowerW: power * 0.85,
      estimatedWh:
          (d2 / 1000) * 14 * (assistWhFactor[climbEst.mode] ?? 1),
      label:
          'Schätzung: ${climbEst.mode.toUpperCase()} (Steigung, ${(climbEst.confidence * 100).round()} %)',
    ),
  );

  final d3 = dist - d1 - d2;
  segments.add(
    AssistSegment(
      id: 'seg-3',
      mode: 'tour',
      source: 'estimated',
      startOffsetSec: t1 + t2,
      endOffsetSec: duration,
      distanceM: d3 < 0 ? 0 : d3,
      avgSpeedKmh: speed,
      avgRiderPowerW: power,
      estimatedWh:
          ((d3 < 0 ? 0 : d3) / 1000) * 9 * (assistWhFactor['tour'] ?? 1),
      label: 'Schätzung: TOUR (Rest)',
    ),
  );

  return summarizeAssist(segments);
}

AssistRideSummary summarizeAssist(List<AssistSegment> segments) {
  final totals = <AssistMode, double>{
    'off': 0,
    'eco': 0,
    'tour': 0,
    'sport': 0,
    'turbo': 0,
  };
  var totalDist = 0.0;
  var wh = 0.0;
  for (final s in segments) {
    totals[s.mode] = (totals[s.mode] ?? 0) + s.distanceM;
    totalDist += s.distanceM;
    wh += s.estimatedWh ?? 0;
  }
  final share = <AssistMode, int>{};
  for (final e in totals.entries) {
    share[e.key] =
        totalDist > 0 ? ((e.value / totalDist) * 100).round() : 0;
  }
  final dominant = totals.entries.toList()
    ..sort((a, b) => b.value.compareTo(a.value));
  return AssistRideSummary(
    segments: segments,
    dominantMode: dominant.isEmpty ? 'tour' : dominant.first.key,
    modeSharePct: share,
    estimatedTotalWh: wh.round(),
    hasEstimates: segments.any((s) => s.source == 'estimated'),
    disclaimer:
        'Schätzungen aus Leistungs-/Geschwindigkeitssignatur — kein OEM-Auslesen. Keine Motorsteuerung (F-EBK-000).',
    sourceLabel:
        'Bosch Battery Guide · BIKE Magazin · Spec F-EBK-005',
  );
}
